liquid-staking/src/delegation.rs

```diff
@@ -0,0 +1,163 @@
+use multiversx_sc::imports::*;
+pub const UNBOND_PERIOD: u64 = 10;
+pub const DELEGATE_ENDPOINT: &str = "delegate";
+pub const UNDELEGATE_ENDPOINT: &str = "unDelegate";
+pub const CLAIM_REWARDS_ENDPOINT: &str = "claimRewards";
+
+use crate::common::{self, storage::Epoch};
+
+#[multiversx_sc::module]
+pub trait DelegationModule: common::storage::CommonStorageModule {
+    #[payable("EGLD")]
+    #[endpoint(stake)]
+    fn stake(&self, contract_name: ManagedBuffer) {
+        let caller = self.blockchain().get_caller();
+        let egld_amount = self.call_value().egld_value().clone_value();
+        let delegation_contract_address = self.delegation_addresses(&contract_name).get();
+
+        self.tx()
+            .to(delegation_contract_address)
+            .raw_call(ManagedBuffer::from(DELEGATE_ENDPOINT))
+            .egld(&egld_amount)
+            .callback(DelegationModule::callbacks(self).stake_callback(&caller, &egld_amount))
+            .async_call_and_exit();
+    }
+
+    #[callback]
+    fn stake_callback(
+        &self,
+        caller: &ManagedAddress,
+        egld_amount: &BigUint,
+        #[call_result] result: ManagedAsyncCallResult<()>,
+    ) {
+        match result {
+            ManagedAsyncCallResult::Ok(()) => {
+                self.delegated_value(caller).set(egld_amount);
+                self.egld_token_supply()
+                    .update(|value| *value += egld_amount)
+            }
+            _ => sc_panic!("There was an error at delegating"),
+        }
+    }
+
+    #[endpoint(unStake)]
+    fn unstake(&self, contract_name: ManagedBuffer, egld_amount_to_unstake: BigUint) {
+        let caller = self.blockchain().get_caller();
+        self.require_caller_has_stake(&caller);
```

Comment:
I'm not sure if you need this validation and the require below, as the delegation SC will throw an error anyway.